graphql/src/store/query.rs

```diff
@@ -360,27 +396,145 @@ fn list_values(value: Value, filter_type: &str) -> Result<Vec<Value>, QueryExecu
     }
 }
 
+enum OrderByValue {
+    Direct(String),
+    Child(String, String),
+}
+
+fn parse_order_by(enum_value: &String) -> OrderByValue {
+    let mut parts = enum_value.split("__");
+    let first = parts.next().expect("some dark magic happened");
+    let second = parts.next();
+
+    match second {
+        Some(second) => OrderByValue::Child(first.to_string(), second.to_string()),
+        None => OrderByValue::Direct(first.to_string()),
+    }
+}
+
+struct ObjectOrderDetails {
+    entity_type: EntityType,
+    join_attribute: Attribute,
+    derived: bool,
+}
+
+struct InterfaceOrderDetails {
+    entity_types: Vec<EntityType>,
+    join_attribute: Attribute,
+    derived: bool,
+}
+
+enum OrderByChild {
+    Object(ObjectOrderDetails),
+    Interface(InterfaceOrderDetails),
+}
+
 /// Parses GraphQL arguments into an field name to order by, if present.
 fn build_order_by(
     entity: ObjectOrInterface,
     field: &a::Field,
-) -> Result<Option<(String, ValueType)>, QueryExecutionError> {
+    schema: &ApiSchema,
+) -> Result<Option<(String, ValueType, Option<OrderByChild>)>, QueryExecutionError> {
     match field.argument_value("orderBy") {
-        Some(r::Value::Enum(name)) => {
-            let field = sast::get_field(entity, name).ok_or_else(|| {
-                QueryExecutionError::EntityFieldError(entity.name().to_owned(), name.clone())
-            })?;
-            sast::get_field_value_type(&field.field_type)
-                .map(|value_type| Some((name.to_owned(), value_type)))
-                .map_err(|_| {
-                    QueryExecutionError::OrderByNotSupportedError(
+        Some(r::Value::Enum(name)) => match parse_order_by(name) {
+            OrderByValue::Direct(name) => {
+                let field = sast::get_field(entity, name.as_str()).ok_or_else(|| {
+                    QueryExecutionError::EntityFieldError(entity.name().to_owned(), name.clone())
+                })?;
+                sast::get_field_value_type(&field.field_type)
+                    .map(|value_type| Some((name.to_owned(), value_type, None)))
+                    .map_err(|_| {
+                        QueryExecutionError::OrderByNotSupportedError(
+                            entity.name().to_owned(),
+                            name.clone(),
+                        )
+                    })
```

Comment:
This is preexisting code, but it also makes me wonder what happens if you manage to use a list or derived field here. I am not super confident that our current validation will catch sorting by such fields, even if they were not in the schema.